Add game and junit tests to AP

SirEndii opened the pull request:
This PR aims to add the long needed game tests to AP. It's still a draft.
We will use the Testing Framework from CC to accomplish this
The framework is implemented, and I started to work on the first tests

- Checklist
- [ ] - Add JUnit tests for common AP functions
- [ ] - Mod Integrations
  - <details>
  
    - [x] - Botania
    - [x] - AE 2
    - [x] - Note Blocks
    - [x] - Beacons
    - [ ] - RS (currently not possible)
    - [ ] - Create
    - [ ] - Mekanism
    - [ ] - Powah
    - [ ] - Minecolonies (currently not possible)
  </details>
- [ ] - Peripherals
  - <details>
  
    - [ ] - Environment Detector
    - [x] - Chat Box
    - [ ] - Player Detector
    - [x] - Energy Detector
    - [ ] - Inventory Manager
    - [x] - Redstone Integrator
    - [x] - Block Reader
    - [x] - Geo Scanner
    - [x] - NBT Storage
  </details>
- [ ] - Turtles
  - <details>
  
    - [ ] - Chatty
    - [ ] - Chunky (Not possible to test)
    - [ ] - Compass
    - [ ] - Player Detector
    - [ ] - Environment Detector
    - [ ] - Geo
    - [ ] - Weak Automata
    - [ ] - OP Weak Automata
    - [ ] - Husbandry Automata
    - [ ] - OP Husbandry Automata
    - [ ] - End Automata
    - [ ] - OP End Automata
  </details>
- [ ] - Pocket Computers (?) Currently not possible without an entity that loads the pocket
  - <details>
  
    - [ ] - Chatty
    - [ ] - Player Detector
    - [ ] - Enviroment
    - [ ] - Geo
  </details>
- [x] - Documentation 


ToDos:
- [ ] The [Geo Scanner docs](https://docs.advanced-peripherals.de/peripherals/geo_scanner/) list a `getScanCooldown` method which currently does not exist in AP, so I cannot test it (the test currently uses the "internal" `getOperationCooldown("scanBlocks")` method)
- [x] The `isOnEnchantedSoil` function on Botania Mana Flowers always returns false, even when the flower is placed on enchanted soil. This seems to be the case because the `overgrowth` variable on a Mana Flower is only set to true during the flower tick (see https://github.com/VazkiiMods/Botania/blob/1.20.x/Xplat/src/main/java/vazkii/botania/api/block_entity/SpecialFlowerBlockEntity.java#L100)
- [ ] The `isEmpty` method on Botania Mana Pools is currently missing
- [ ] For the Botania Mod Integration, some methods are only available in the 1.20.1 versions of AP, so they're currently commented out in the .lua test files. When merging this to the 1.20.1 branch, these methods should be uncommented
- [ ] Botania Mana Spreaders have a `getBounding` method returning the coordinates of the block that the spreader is directed towards, shouldn't it be called `getBinding` then?
- [ ] The `isEmpty` function on Botania Mana Spreaders does not return if the mana in the spreader is empty but if the item handler of the spreader is empty
- [ ] On Note Blocks, the method `changeNoteBy` is named somewhat weird. It sets the note value to the given parameter and does not change the note by x amount. Shouldn't it be called `setNote` or something?

SirEndii committed:
Start writing some new tests, this not even near finished. The plan is to code test the most code which does not need a running minecraft instance. For other parts of the project, gametests will be created at a later point

Also make a static helper function to create the tooltips to minimize duplication and maintenance time

SirEndii committed:
Merge branch '1.19.2' into dev/tests

SirEndii committed:
Merge branch '1.19.2' into dev/tests

SirEndii committed:
Merge branch '1.19.2' into dev/tests

SirEndii committed:
Merge branch '1.19.2' into dev/tests

# Conflicts:
#	gradle.properties

SirEndii committed:
Merge branch '1.19.2' into dev/tests

SirEndii committed:
Merge branch 'refs/heads/dev/1.19.2' into feat/tests

# Conflicts:
#	src/main/java/de/srendi/advancedperipherals/common/items/base/BaseBlockItem.java
#	src/main/java/de/srendi/advancedperipherals/common/items/base/BaseItem.java

SirEndii committed:
Add some testing game tests, next step is trying to run tests on computers with CC's testing api
Removed some old stuff

SirEndii committed:
Remove unused imports

SirEndii committed:
Add the CC testing framework utilities

SirEndii committed:
Merge branch 'refs/heads/dev/1.19.2' into feat/tests

# Conflicts:
#	build.gradle

SirEndii committed:
Merge branch 'refs/heads/dev/1.19.2' into feat/tests

SirEndii commented:
I will mark this as high priority. I will work on this asap since we really need tests for AP

SirEndii committed:
Merge branch 'refs/heads/dev/1.19.2' into feat/tests

SirEndii committed:
Implemented(Copied) the needed classes for the computer game tests from the CC repository and adapted it to our needs

SirEndii commented:
I've implemented (copied) the needed stuff from the test source sets, but there is currently another issue with hamcrest and junit I need to work on
I'm going to fix this asap. For now there are the cc tests in AP for testing purposes

SirEndii committed:
Added needed junit and hamcrest libraries to the minecraft classpath so they are included in the runtime when we run the tests

SirEndii committed:
Added jqwik database to gitignore

SirEndii committed:
Fixed game tests for `runGameTestClient`

SirEndii commented:
I fixed everything that need to be fixed (I think)
The game tests are now ready to be created

zyxkad commented on build.gradle:
```diff
@@ -126,38 +148,69 @@ minecraft {
+                def allFiles = new HashSet<String>();
+
+                def oldVal = old?.get()
+                if (!oldVal != null && !oldVal.isEmpty()) allFiles.addAll(oldVal.split(File.pathSeparatorChar))
```
what is `!oldVal != null`

SirEndii replied:
That's the minecraft classpath
The value of the returned supplier can be null, so we check that

zyxkad commented on src/testMod/java/de/srendi/advancedperipheralstest/TestGameTests.java:
```diff
@@ -0,0 +1,19 @@
+package de.srendi.advancedperipheralstest;
```
```suggestion
package de.srendi.advancedperipherals.test;
```

zyxkad commented on src/testMod/kotlin/dan200/computercraft/gametest/Computer_Test.kt:
```diff
@@ -0,0 +1,72 @@
+/*
```
Is these gametests from CC:T really necessary? Is our test depending on it?

zyxkad commented on src/testMod/resources/data/advancedperipheralstest/computer/tests/craftos_test.sends_basic_rednet_messages.echo.lua:
```diff
@@ -0,0 +1,8 @@
+os.getComputerID = function() return 1 end
```
I don't think we need to test CC:T stuffs in AP.

SirEndii replied:
Yeah these are just for testing. I used them to see if everything works fine in the testing environment

SirEndii committed:
Rename var old -> oldClasspath, oldVal -> minecraftClasspath

SirEndii committed:
Remove some not needed stuff in the build.gradle

SirEndii committed:
Created a test using kotlin for the environmental detector. This was just a test, there is a bug loading the structures what I need to fix.

zyxkad commented on build.gradle:
```diff
@@ -126,38 +148,69 @@ minecraft {
+                def allFiles = new HashSet<String>();
+
+                def minecraftClasspath = oldClasspath?.get()
+                if (!minecraftClasspath != null && !minecraftClasspath.isEmpty()) allFiles.addAll(minecraftClasspath.split(File.pathSeparatorChar))
```
Well, I mean, is **`!`** `minecraftClasspath != null` a correct syntax? I checked gradle docs and the syntax should same as java?

SirEndii replied:
Ahhh that's why intellij showed me a warning 
Didn't saw that, oops

SirEndii committed:
Remove CC tests and complete the environment detector test
The test can be run via `runGameTestClient` using the /test command or `runGameTestServer` to run all the tests automatically

SirEndii committed:
Add game test action

SirEndii committed:
Satisfy checkstyle

SirEndii committed:
Fixed if statement

SirEndii commented:
We're on a good run.
I've added the first game tests using an ingame script. I will see if I can run the functions via the kotlin script like CC does it with the APIs like turtle.*. But I doubt that this is easily possible with peripherals

We just now need to add the game tests itself. I will also write a documentation for that

SirEndii commented:
Had another error in my thinking. gradle build already executes test and runGameTestServer, I'll change it later

SirEndii committed:
Added a prototype test for the ME Bridge. Used to know how the ME System handles the game tests. We need to wait a bit using `sleep` in the test lua script so the system has enough time to boot up

Don't implement kotlinforforge to prevent issues with kotlin and kotlin coroutines

SirEndii committed:
Added documentation on how to create tests.
This is not completely done, but gives a first impression on how to create tests and how to run them

SirEndii commented:
I think there is nothing left to do in terms of the framework. Now we just need to create the tests. There is a me bridge tests to test something a bit more complex and the environment detector which currently just tests one function.

I've created a documentation markdown file located in docs/CREATING_TESTS.md. It explains the basics, how to run and write tests. But it needs some more love with examples and code.
If someone wants to create tests, feel free to do so😅  Would speed things up
For non collaborators, you could create a PR which targets the feat/tests branch. But I am not too sure how well this goes

zyxkad commented on src/testMod/resources/data/advancedperipheralstest/computer/tests/peripheraltest.environment.lua:
```diff
@@ -0,0 +1,5 @@
+detector = peripheral.find("environmentDetector")
+test.eq(detector ~= nil, true, "Peripheral not found")
```
not sure if the method exists but I believe `test.neq(detector, nil` will better than `test.eq(x ~= nil`

zyxkad commented on src/testMod/resources/data/advancedperipheralstest/computer/tests/peripheraltest.environment.lua:
```diff
@@ -0,0 +1,5 @@
+test.eq(detector ~= nil, true, "Peripheral not found")
+
+isRaining = detector.isRaining()
+test.eq(false,isRaining, "It should not rain")
```
```suggestion
test.eq(false, isRaining, "It should not rain")
```

zyxkad commented on src/testMod/resources/data/advancedperipheralstest/computer/tests/peripheraltest.mecrafting.lua:
```diff
@@ -0,0 +1,30 @@
+test.assert(isOnline, "Bridge is not connected/system is not online")
+
+validEnergy = bridge.getEnergyUsage() > 24
+test.assert(validEnergy, tostring(bridge.getEnergyUsage()) .. "Consumption does not seem right")
```
```suggestion
test.assert(validEnergy, tostring(bridge.getEnergyUsage()) .. " Consumption does not seem right")
```

or `string.format` will looks better than string join

zyxkad commented on src/testMod/resources/data/advancedperipheralstest/computer/tests/peripheraltest.mecrafting.lua:
```diff
@@ -0,0 +1,30 @@
+-- Logs should not be craftable, we don't have a pattern for it
+isItemCraftable = bridge.getItem(logFilter).isCraftable
+-- We use == false here since the not keyword would also return true if the value is nil
+test.assert(isItemCraftable == false, "Log should not be craftable")
```
```suggestion
test.assert(not isItemCraftable, "Log should not be craftable")
```

SirEndii replied:
There is a comment above that line which explains why I did that

zyxkad commented on src/testMod/resources/data/advancedperipheralstest/computer/tests/peripheraltest.mecrafting.lua:
```diff
@@ -0,0 +1,30 @@
+
+-- Planks are craftable
+isItemCraftable = bridge.getItem(planksFilter).isCraftable
+test.assert(isItemCraftable == true, "Planks should be craftable")
```
```suggestion
test.assert(isItemCraftable, "Planks should be craftable")
```

it's better also keep same style when writing lua scripts, either all of them use `xxx == true`, `xxx == false` or all of them use `xxx`, `not xxx`

SirEndii replied:
Yeah you're right
But sometimes it does not really make sense to use == true/false
And sometimes I just need to use == true/false like in line 25

SirEndii committed:
Some styling changes to the lua scripts

SirEndii committed:
Polished andesite

SirEndii committed:
eq -> assert and space fixing in the docs

SirEndii committed:
Finished the me crafting test and started implementing a test for me storage related function where I need to fix a bug in AP first

SirEndii committed:
Merge branch 'refs/heads/dev/1.19.2' into feat/tests

SirEndii committed:
Finished the me storage test

SirEndii committed:
Added me transfer tests to test the integrity of our item export and import tests

dogboy21 committed:
Add game tests for Block Reader peripheral

dogboy21 committed:
Fix inconsistency in the code style for assert/eq on isTileEntity

dogboy21 committed:
tablelength -> tableLength

dogboy21 committed:
Implement tests for NBT Storage peripheral

SirEndii committed:
Merge pull request #610 from dogboy21/blockreader-tests

Add game tests for Block Reader peripheral (related to #562)

dogboy21 committed:
Implement tests for Redstone Integrator peripheral

dogboy21 committed:
Implement tests for Geo Scanner peripheral

dogboy21 committed:
Add missing comment in Geo Scanner test

dogboy21 committed:
Add test for Mana Flowers in the Botania Mod Integration

dogboy21 committed:
Add todo comment for missing getScanCooldown function on Geo Scanner

dogboy21 committed:
Uncomment tests for currently unavailable Botania methods

dogboy21 committed:
Add test for Mana Pools in the Botania Mod Integration

dogboy21 committed:
Add test for Mana Spreaders in the Botania Mod Integration

dogboy21 committed:
Add test for Beacons in the Minecraft Mod (lol) Integration

dogboy21 committed:
Add test for Note Blocks in the Minecraft Mod Integration

dogboy21 committed:
Fix style for object key accessors

dogboy21 committed:
Fix RS Integrator test flakiness

SirEndii committed:
Merge branch 'refs/heads/dev/1.19.2' into feat/tests

SirEndii committed:
Merge remote-tracking branch 'refs/remotes/origin/dev/1.19.2' into feat/tests

SirEndii committed:
increase sleep for the `isItemCrafting` function in the mecrafting peripheral test

dogboy21 committed:
Readd test for isOnEnchantedSoil on Botania Flowers now that the method works correctly

SirEndii committed:
Merge pull request #612 from dogboy21/peripheral-tests

Implement game tests for more peripherals

SirEndii committed:
Revert "Implement game tests for more peripherals"

SirEndii committed:
Merge pull request #615 from IntelligenceModding/revert-612-peripheral-tests

Revert "Implement game tests for more peripherals"

SirEndii committed:
Revert "Revert "Implement game tests for more peripherals""

SirEndii committed:
Merge pull request #616 from IntelligenceModding/revert-615-revert-612-peripheral-tests

Revert "Revert "Implement game tests for more peripherals""

SirEndii committed:
Some documentation additions

SirEndii committed:
Fixed the orientation of RS Blocks when they are loaded from a structure block. #BlameRS

SirEndii committed:
Begone

dogboy21 committed:
Backported changes from the CC: Tweaked Gametest framework present in the 1.20.x branch

This was mainly done to allow the execution of client tests

dogboy21 committed:
Add gradle task to start automated client tests

dogboy21 committed:
Add docs on client tests

SirEndii committed:
Merge pull request #622 from dogboy21/cctest-backport

Backported changes from the CC: Tweaked Gametest framework present in 1.20.x

dogboy21 committed:
Added client test for the Chat Box peripheral

dogboy21 committed:
Fix NPE on client test start with existing world

dogboy21 committed:
Add Chat Box event test

dogboy21 committed:
Fix runGameTestClient task also running common tests

dogboy21 committed:
Clean and fix runGameTestClient task

SirEndii committed:
Merge pull request #628 from dogboy21/clienttests

Added client test for the Chat Box peripheral

SirEndii committed:
Energy Detector tests

SirEndii committed:
Merge branch 'dev/0.8' into feat/tests

# Conflicts:
#	build.gradle
#	gradle.properties
#	src/main/java/de/srendi/advancedperipherals/common/util/inventory/ItemUtil.java

SirEndii committed:
snake case peripheral types, some ae2 test fixes

SirEndii committed:
Merge branch 'dev/0.8' into feat/tests

zyxkad commented:
by the way why is chunky turtle not possible to test?
I think the test cases are:

- load when server start:
   1. place turtle to (edge of/far away in) the world
   2. restart server
   3. check if the turtle id exists, and/or check if the chunk is loaded
- unload chunk when moving/destroyed:
   1. place turtle to far away in the world
   2. make the turtle move across the border, or destroy the turtle
   3. check if the back chunk is unloaded
- load chunk when moving:
   1. place turtle to far away in the world
   2. make the turtle move across the border
   3. check if the front/side chunk is loaded

I do not know if that's possible to do with the current test framework.

SirEndii committed:
Adapted some me bridge tests to the new me bridge peripheral

SirEndii commented:
Okay you are right, not very complicated

SirEndii committed:
Using powah for the energy detector test to prevent a crashing NPE with mekanism
Fixed the me transfer test

SirEndii committed:
increase the timeoutticks for the energy detector test

SirEndii committed:
Test our `isFluidCrafting` and `getFluid` function

SirEndii committed:
Mekanism Environment detector radiation test

SirEndii commented:
I currently just tend to merge this, even unfinished. Imo testing is important, especially with all the new changes. But I don't have the time to implement ALL of that.
I will change my to-do list a bit and sort it by priority. Something which is annoying to test by hand should have the highest prio and vice versa
